Firestore/core/test/firebase/firestore/local/index_free_query_engine_test.cc

```diff
@@ -0,0 +1,336 @@
+
+SnapshotVersion LAST_LIMBO_FREE_SNAPSHOT = Version(10);
+SnapshotVersion MISSING_LAST_LIMBO_FREE_SNAPSHOT = SnapshotVersion::None();
+
+}  // namespace
+
+class TestLocalDocumentsView : public LocalDocumentsView {
+ public:
+  TestLocalDocumentsView(RemoteDocumentCache* remote_document_cache,
+                         MutationQueue* mutation_queue,
+                         IndexManager* index_manager)
+      : LocalDocumentsView(
+            remote_document_cache, mutation_queue, index_manager) {
+  }
+
+  model::DocumentMap GetDocumentsMatchingQuery(
+      const core::Query& query,
+      const model::SnapshotVersion& since_read_time) override {
+    EXPECT_TRUE(expect_index_free_execution.has_value());
+    EXPECT_EQ(*expect_index_free_execution,
+              since_read_time != SnapshotVersion::None());
+    return LocalDocumentsView::GetDocumentsMatchingQuery(query,
+                                                         since_read_time);
+  }
+
+  void ExpectIndexFreeExecution(bool index_free) {
+    expect_index_free_execution = index_free;
+  }
+
+ private:
+  absl::optional<bool> expect_index_free_execution;
+};
+
+class IndexFreeQueryEngineTest : public ::testing::Test {
+ public:
+  IndexFreeQueryEngineTest()
+      : persistence_(MemoryPersistence::WithEagerGarbageCollector()),
+        remote_document_cache_(persistence_->remote_document_cache()),
+        query_cache_(persistence_->query_cache()),
+        index_manager_(new MemoryIndexManager()),
+        local_documents_view_(
+            remote_document_cache_,
+            persistence_->GetMutationQueueForUser(User::Unauthenticated()),
+            index_manager_.get()) {
+    query_engine_.SetLocalDocumentsView(&local_documents_view_);
+  }
+
+  /** Adds the provided documents to the query target mapping. */
+  void PersistQueryMapping(const std::vector<DocumentKey>& keys) {
+    persistence_->Run("PersistQueryMapping", [this, &keys]() {
+      DocumentKeySet remote_keys;
+      for (const DocumentKey& key : keys) {
+        remote_keys = remote_keys.insert(key);
+      }
+      query_cache_->AddMatchingKeys(remote_keys, TEST_TARGET_ID);
+    });
+  }
+
+  /** Adds the provided documents to the remote document cache. */
+  void AddDocuments(const std::vector<Document>& docs) {
+    persistence_->Run("AddDocuments", [this, &docs]() {
+      for (const Document& doc : docs) {
+        remote_document_cache_->Add(doc, doc.version());
+      }
+    });
+  }
+
+  DocumentSet ExpectIndexFreeQuery(const std::function<DocumentSet(void)>& f) {
+    local_documents_view_.ExpectIndexFreeExecution(true);
+    return f();
+  }
+
+  DocumentSet ExpectFullCollectionQuery(
+      const std::function<DocumentSet(void)>& f) {
+    local_documents_view_.ExpectIndexFreeExecution(false);
+    return f();
+  }
+
+  DocumentSet RunQuery(
+      const Query& query,
+      const SnapshotVersion& last_limbo_free_snapshot_version) {
+    const DocumentKeySet& remote_keys =
```

Comment:
Hrm. I am guessing I need to modify my "always assign to a reference" rule that I had just made up. But what you are explaining seems reasonable. So, basically, if the method returns a const reference, I should assign to one, but if it returns an rvalue I shouldn't since I don't know anything about the lifetime guarantees? 

Fixed.

Reply:
So in this case, I think it's theoretically possible to come up with a deterministic flowchart of which one to pick, though it would be pretty involved because there's several things to consider (value categories (rvalue vs lvalue), built-in vs user-defined types, lifetime extension). I'll try a couple of rationales in case you find any of it useful.

One way to look at it is that a reference can be seen as creating another name for an object. When the object is passed as an argument, using a reference is the only way the syntax allows for referring to the same object (ignoring pointers for the purposes of this discussion). When, however, we have access to the object in question within our function, using a reference is mostly a matter of convenience. It can be seen as an alias then; it's just more readable (and easier to write) to use `key` than `kv.second`, for example.

On the other hand, when we're dealing with a temporary, saying that we give it another, more convenient name doesn't really make sense because a temporary has no name in the first place. That's one way to look at this that perhaps can illustrate why a reference is less preferable in this case.

Another way to see this is that we're using references to avoid unnecessary copies. A copy is unnecessary when we simply want to read some fields from an object, so that's why using a const reference in this case:
```cpp
for (const DocumentKey& key : set) {
  // We only read from `key`
```
is preferable to creating a copy:
```cpp
for (DocumentKey key: set) {
  // We only read from `key`, so we don't really need a copy of it
```

In the case of a temporary, however, there's no copy to avoid, so using a reference doesn't bring any performance improvement. This is because of named return value optimization ([NRVO](https://en.cppreference.com/w/cpp/language/copy_elision#Non-mandatory_elision_of_copy.2Fmove_.28since_C.2B.2B11.29_operations)):
```cpp
DocumentKey key = CreateKey(); // No copy takes place due to NRVO
const DocumentKey& key = CreateKey(); // Also no copy
```

NRVO makes it so that when a local variable is returned from a function, and that function's return value is used to initialize (not reassign, initialize) a variable, no copy takes place (the more specific term is that the copy is _elided_). There are some limitations to NRVO, but I wouldn't be concerned about them generally.

> So, basically, if the method returns a const reference, I should assign to one, but if it returns an rvalue I shouldn't since I don't know anything about the lifetime guarantees?

Yes, except that you do know about the lifetime guarantee -- due to the way const references extend the lifetime of a temporary they're bound to, the temporary will stay alive as long as the reference is alive (AFAIK, safety was the major reason this feature was added). I'd just generally prefer not to rely on this behavior, especially in non-generic code. Using a plain variable is just easier and more straightforward for storing the temporary, with no downsides I can think of.